### PR TITLE
render `rule.meta.mbc` on output

### LINKED ISSUE #526: parse MBC technique into fields for the JSON output format
> structuring it, like the rest of the JSON output e.g. {"ID":"B0001.008","tactic":"Anti-Behavioral Analysis","technique":"Debugger Detection","subtechnique":"IsDebuggerPresent"}, following https://raw.githubusercontent.com/MBCProject/mbc-markdown/master/yfaq/mbc_matrix_with_ids.svg

> parsing the string for the json output is a good idea - it reduces the number of times that programmatic users of capa have to write one-off parsers for the MBC information.

from https://github.com/fireeye/capa-rules/issues/316

### PR DESCRIPTION
**Pull request by mr-tz**
### Description

I think a better approach than #574 

closes #526 

### Documentation

- [ ] I have updated the [CHANGELOG.md](/CHANGELOG.md), this is required for:
  - Bug fixes (non-breaking change which fixes an issue)
  - New features (non-breaking change which adds functionality)
  - Breaking changes (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
  - [ ] I have made the corresponding changes to the documentation

### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] No new tests needed

### CODE REVIEW

**Review by williballenthin — CHANGES_REQUESTED** [review]
see comments in #574

File: capa/render/default.py [hunk 1]
```diff
@@ -179,30 +179,22 @@ def render_mbc(doc, ostream):
         if not rule["meta"].get("mbc"):
             continue
 
-        mbcs = rule["meta"]["mbc"]
-        if not isinstance(mbcs, list):
-            raise ValueError("invalid rule: MBC mapping is not a list")
```
Comment (mr-tz) [inline capa/render/default.py:184]: do this check now when loading the rule, see below

File: capa/render/__init__.py [hunk 2]
```diff
@@ -162,6 +162,40 @@ def convert_match_to_result_document(rules, capabilities, result):
     return doc
 
 
+def convert_meta_to_result_document(meta):
+    mbcs = meta.get("mbc", [])
+    meta["mbc"] = [parse_canonical_mbc(mbc) for mbc in mbcs]
+    return dict(meta)
```
Comment (Ana06) [inline capa/render/__init__.py:168]: isn't `meta` already a `dict`? 😕

File: capa/render/__init__.py [hunk 3]
```diff
@@ -162,6 +162,40 @@ def convert_match_to_result_document(rules, capabilities, result):
...
+        objective = parts[0]
+    if len(parts) > 1:
+        behavior = parts[1]
+    if len(parts) > 2:
+        method = parts[2]
+
+    return {
+        "parts": parts,
+        "id": id or "",
+        "objective": objective or "",
```
Comment (Ana06) [inline capa/render/__init__.py:193]: We could initialize `objective` to `""` instead `None` and then this `or` is not needed.

**Review by Ana06 — COMMENTED** [review]
This deserves a CHANGELOG entry 😉

File: capa/render/__init__.py (hunk 2 above)
Comment (williballenthin) [reply capa/render/__init__.py:168]: i *think* it might also be a ruamel/yaml dict-like object, let me confirm.

Comment (williballenthin) [reply capa/render/__init__.py:168]: im not exactly sure. i agree, if no tests complain when removing this cast, then we should prefer the simpler code.

File: capa/render/__init__.py [hunk 4]
```diff
@@ -162,6 +162,40 @@ def convert_match_to_result_document(rules, capabilities, result):
...
+    mbcs = meta.get("mbc", [])
+    meta["mbc"] = [parse_canonical_mbc(mbc) for mbc in mbcs]
+    return dict(meta)
+
+
+def parse_canonical_mbc(mbc):
+    """
+    parse capa's canonical MBC representation: `Objective::Behavior::Method [Identifier]`
+    """
+    id = objective = behavior = method = None
```
Comment (williballenthin) [inline capa/render/__init__.py:175]:
```suggestion
    id = None
    objective = None
    behavior = None
    method = None
```

File: capa/render/utils.py [hunk 5]
```diff
@@ -29,6 +29,10 @@ def hex(n):
         return "0x%X" % n
 
 
+def format_mbc(mbc):
+    return "%s [%s]" % ("::".join(mbc["parts"]), mbc["id"])
```
Comment (williballenthin) [inline capa/render/utils.py:33]: 👍🏼

**Review by williballenthin — APPROVED** [review]

File: capa/render/__init__.py [hunk 6]
```diff
@@ -162,6 +162,43 @@ def convert_match_to_result_document(rules, capabilities, result):
     return doc
 
 
+def convert_meta_to_result_document(meta):
+    mbcs = meta.get("mbc", [])
+    meta["mbc"] = [parse_canonical_mbc(mbc) for mbc in mbcs]
+    return meta
```
Comment (mr-tz) [inline capa/render/__init__.py:168]: confirmed that this is already a dict

File: CHANGELOG.md [hunk 7]
```diff
@@ -111,6 +111,7 @@ It includes many new rules, including all new techniques introduced in MITRE ATT
 - linter: summarize results at the end #571 @williballenthin
 - meta: added `library_functions` field, `feature_counts.functions` does not include library functions any more #562 @mr-tz
 - linter: check for `or` with always true child statement, e.g. `optional`, colors #348 @mr-tz
+- json: breaking change in results document; now contains parsed MBC fields instead of canonical representation #526 @mr-tz
```
Comment (williballenthin) [inline CHANGELOG.md:114]: i think this isnt a breaking change: we add a new field, but don't change or remove and old field? or did the type of the existing field change?

**Pull request merged**